File: backend/app/schemas/molecule_admin.py
"""Request/response models for the Molecule Data admin page.

``MoleculeDraft`` mirrors the shape of an entry in ``curated_molecules.json``
(see that file for real examples) plus a handful of additional teaching and
provenance fields the admin UI exposes that the baseline file does not yet
use. ``extra="allow"`` keeps it forward-compatible with fields this module
does not need to know about, while the declared fields give the editor UI and
OpenAPI schema a concrete shape to build against.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field

from app.properties.schema import NormalizedProperty
from app.schemas.geometry_evidence_schema import MolecularGeometryEvidence


class AdminLoginRequest(BaseModel):
    username: str = Field(min_length=1, max_length=80)
    password: str = Field(min_length=1, max_length=200)


class AdminSessionStatus(BaseModel):
    authenticated: bool
    username: str | None = None


class ExceptionFlags(BaseModel):
    electron_deficient: bool = False
    expanded_octet: bool = False
    odd_electron: bool = False


class ThreeDSource(BaseModel):
    kind: str = "idealized_vsepr_template"
    verified_reference: bool = False


class ResonanceStructureDraft(BaseModel):
    form_index: int
    bond_orders: list[int]
    lone_pairs: list[int]
    formal_charges: list[int]


class ReviewProvenance(BaseModel):
    """The Source & Review tab's own provenance block, separate from per-observation
    geometry/property sources -- this is the record-level "who curated this" trail."""

    source_name: str | None = None
    reference: str | None = None
    url: str | None = None
    evidence_type: str | None = None
    conditions: str | None = None
    retrieved_at: datetime | None = None


class MoleculeDraft(BaseModel):
    model_config = ConfigDict(extra="allow")

    id: str = Field(min_length=1, max_length=80)
    formula: str = Field(min_length=1, max_length=80)
    charge: int = 0
    name_vi: str = ""
    name_en: str = ""
    aliases: list[str] = Field(default_factory=list)

    cas_rn: str | None = None
    pubchem_cid: int | None = None
    smiles: str | None = None
    inchi: str | None = None
    inchikey: str | None = None

    atom_inventory: dict[str, int] = Field(default_factory=dict)
    atom_symbols: list[str] = Field(default_factory=list)
    central_atom: str = ""

    total_valence_electrons: int = 0
    bond_orders: list[int] = Field(default_factory=list)
    lone_pairs: list[int] = Field(default_factory=list)
    formal_charges: list[int] = Field(default_factory=list)
    resonance_forms: int = 1
    resonance_structures: list[ResonanceStructureDraft] = Field(default_factory=list)
    resonance_note_vi: str | None = None
    resonance_note_en: str | None = None
    exception_flags: ExceptionFlags = Field(default_factory=ExceptionFlags)

    bonding_domains: int = 0
    lone_pair_domains: int = 0
    steric_number: int = 0
    ax_en: str = ""
    electron_geometry: str = ""
    electron_geometry_vi: str = ""
    molecular_geometry: str = ""
    molecular_geometry_vi: str = ""
    ideal_angle: str = ""
    distortion_note_vi: str | None = None
    distortion_note_en: str | None = None
    hybridization: str | None = None
    hybridization_warning_vi: str | None = None
    hybridization_warning_en: str | None = None
    polarity_note_vi: str | None = None
    polarity_note_en: str | None = None

    teaching_note_vi: str | None = None
    teaching_note_en: str | None = None
    misconception_note_vi: str | None = None
    misconception_note_en: str | None = None
    structure_property_note_vi: str | None = None
    structure_property_note_en: str | None = None

    three_d_source: ThreeDSource = Field(default_factory=ThreeDSource)
    source: Literal["curated", "cache", "PubChem reference", "deterministic"] = "curated"
    confidence: Literal["high", "medium", "low"] = "medium"
    review_status: str = "draft"
    review_provenance: ReviewProvenance | None = None


class MoleculeAdminSaveRequest(BaseModel):
    molecule: MoleculeDraft
    experimental_geometry: MolecularGeometryEvidence | None = None
    properties: list[NormalizedProperty] = Field(default_factory=list)


class MoleculeAdminRecord(BaseModel):
    molecule: dict[str, Any]
    experimental_geometry: MolecularGeometryEvidence | None = None
    properties: list[NormalizedProperty] = Field(default_factory=list)
    has_override: bool = False
    is_admin_added: bool = False


class MoleculeAdminListItem(BaseModel):
    id: str
    formula: str
    charge: int
    name_vi: str
    name_en: str
    ax_en: str
    molecular_geometry: str
    molecular_geometry_vi: str
    review_status: str
    has_override: bool = False
    is_admin_added: bool = False


class MoleculeAdminListResponse(BaseModel):
    results: list[MoleculeAdminListItem]


class ValidationIssue(BaseModel):
    severity: Literal["error", "warning", "info"]
    field: str | None = None
    message_vi: str
    message_en: str


class ValidationReport(BaseModel):
    is_valid: bool
    errors: list[ValidationIssue] = Field(default_factory=list)
    warnings: list[ValidationIssue] = Field(default_factory=list)
    info: list[ValidationIssue] = Field(default_factory=list)


class MoleculeAdminSaveResponse(BaseModel):
    molecule: MoleculeAdminListItem
    validation: ValidationReport
    saved_at: datetime


class CompletenessReport(BaseModel):
    molecule_id: str
    missing_fields: list[str] = Field(default_factory=list)
    has_experimental_geometry: bool = False
    has_properties: bool = False
    review_status: str = ""
    completeness_percent: float = 0.0


class DraftGenerationRequest(BaseModel):
    formula: str = Field(min_length=1, max_length=80)
    charge: int = 0
    id: str | None = None
    name_vi: str | None = None
    name_en: str | None = None

File: frontend/src/types/moleculeAdmin.ts
/** Mirrors `backend/app/schemas/molecule_admin.py`. */

import type { NormalizedProperty } from "./properties";

export interface AdminSessionStatus {
  authenticated: boolean;
  username: string | null;
}

export interface ExceptionFlags {
  electron_deficient: boolean;
  expanded_octet: boolean;
  odd_electron: boolean;
}

export interface ThreeDSource {
  kind: string;
  verified_reference: boolean;
}

export interface ResonanceStructureDraft {
  form_index: number;
  bond_orders: number[];
  lone_pairs: number[];
  formal_charges: number[];
}

export interface ReviewProvenance {
  source_name: string | null;
  reference: string | null;
  url: string | null;
  evidence_type: string | null;
  conditions: string | null;
  retrieved_at: string | null;
}

/** A molecule record, editable through the admin page. Extra fields the frontend
 * does not know about round-trip unchanged (the backend model allows them). */
export interface MoleculeDraft {
  id: string;
  formula: string;
  charge: number;
  name_vi: string;
  name_en: string;
  aliases: string[];

  cas_rn: string | null;
  pubchem_cid: number | null;
  smiles: string | null;
  inchi: string | null;
  inchikey: string | null;

  atom_inventory: Record<string, number>;
  atom_symbols: string[];
  central_atom: string;

  total_valence_electrons: number;
  bond_orders: number[];
  lone_pairs: number[];
  formal_charges: number[];
  resonance_forms: number;
  resonance_structures: ResonanceStructureDraft[];
  resonance_note_vi: string | null;
  resonance_note_en: string | null;
  exception_flags: ExceptionFlags;

  bonding_domains: number;
  lone_pair_domains: number;
  steric_number: number;
  ax_en: string;
  electron_geometry: string;
  electron_geometry_vi: string;
  molecular_geometry: string;
  molecular_geometry_vi: string;
  ideal_angle: string;
  distortion_note_vi: string | null;
  distortion_note_en: string | null;
  hybridization: string | null;
  hybridization_warning_vi: string | null;
  hybridization_warning_en: string | null;
  polarity_note_vi: string | null;
  polarity_note_en: string | null;

  teaching_note_vi: string | null;
  teaching_note_en: string | null;
  misconception_note_vi: string | null;
  misconception_note_en: string | null;
  structure_property_note_vi: string | null;
  structure_property_note_en: string | null;

  three_d_source: ThreeDSource;
  source: "curated" | "cache" | "PubChem reference" | "deterministic";
  confidence: "high" | "medium" | "low";
  review_status: string;
  review_provenance: ReviewProvenance | null;
}

export interface GeometryAtomDraft { id: string; element: string; role: "center" | "ligand" | "other" }
export interface GeometryBondDraft { atom1_id: string; atom2_id: string; order: number }
export interface GeometryCoordinateDraft { id: string; element: string; x: number; y: number; z: number }
export interface GeometryObservationSourceDraft {
  source_name: string | null; source_reference: string | null; source_url: string | null;
  comment: string | null; retrieval_timestamp: string | null;
}
export interface BondLengthObservationDraft {
  id: string; atom1_id: string; atom2_id: string; value_angstrom: number;
  uncertainty_angstrom: number | null; equivalent_count: number; label: string | null;
  source: GeometryObservationSourceDraft | null;
}
export interface BondAngleObservationDraft {
  id: string; atom1_id: string; center_atom_id: string; atom2_id: string; value_deg: number;
  uncertainty_deg: number | null; equivalent_count: number; label: string | null;
  source: GeometryObservationSourceDraft | null;
}
export interface DihedralObservationDraft {
  id: string; atom1_id: string; atom2_id: string; atom3_id: string; atom4_id: string; value_deg: number;
  uncertainty_deg: number | null; equivalent_count: number; label: string | null;
  source: GeometryObservationSourceDraft | null;
}
export interface GeometryIdentityDraft {
  formula: string; charge: number; atom_inventory: Record<string, number>;
  inchi: string | null; inchikey: string | null; cas_rn: string | null; pubchem_cid: number | null;
  canonical_identity: string | null; curated_molecule_id: string | null; formula_identity_unambiguous: boolean;
}
export interface GeometrySourceDraft {
  name: string; reference: string | null; url: string | null; comments: string | null; retrieved_at: string | null;
}

export type GeometryEvidenceKind = "experimental" | "source_annotation" | "computed_conformer" | "deterministic_calculation" | "ideal_vsepr";

export interface MolecularGeometryEvidenceDraft {
  id: string;
  identity: GeometryIdentityDraft;
  evidence_type: GeometryEvidenceKind;
  atoms: GeometryAtomDraft[];
  bonds: GeometryBondDraft[];
  bond_lengths: BondLengthObservationDraft[];
  bond_angles: BondAngleObservationDraft[];
  dihedrals: DihedralObservationDraft[];
  coordinates: GeometryCoordinateDraft[] | null;
  units: "angstrom";
  phase: string | null;
  electronic_state: string | null;
  conformation: string | null;
  point_group: string | null;
  source: GeometrySourceDraft;
}

export interface MoleculeAdminListItem {
  id: string;
  formula: string;
  charge: number;
  name_vi: string;
  name_en: string;
  ax_en: string;
  molecular_geometry: string;
  molecular_geometry_vi: string;
  review_status: string;
  has_override: boolean;
  is_admin_added: boolean;
}

export interface MoleculeAdminRecord {
  molecule: MoleculeDraft;
  experimental_geometry: MolecularGeometryEvidenceDraft | null;
  properties: NormalizedProperty[];
  has_override: boolean;
  is_admin_added: boolean;
}

export interface ValidationIssue {
  severity: "error" | "warning" | "info";
  field: string | null;
  message_vi: string;
  message_en: string;
}

export interface ValidationReport {
  is_valid: boolean;
  errors: ValidationIssue[];
  warnings: ValidationIssue[];
  info: ValidationIssue[];
}

export interface MoleculeAdminSaveResponse {
  molecule: MoleculeAdminListItem;
  validation: ValidationReport;
  saved_at: string;
}

export interface CompletenessReport {
  molecule_id: string;
  missing_fields: string[];
  has_experimental_geometry: boolean;
  has_properties: boolean;
  review_status: string;
  completeness_percent: number;
}

export interface MoleculeAdminSavePayload {
  molecule: MoleculeDraft;
  experimental_geometry?: MolecularGeometryEvidenceDraft | null;
  properties?: NormalizedProperty[];
}
